Sources/ViewType.swift

```diff
@@ -0,0 +1,225 @@
+//
+
+import UIKit
+
+private var layoutAssociationKey = "layout_association_key"
+
+public enum ViewType {
+    case view(UIView)
+    case layer(CALayer)
+}
+
+extension ViewType {
+
+    var layout: Layout {
+        if let value = objc_getAssociatedObject(self, &layoutAssociationKey) as? Layout {
+            return value
+        }
+        else {
+            let layout: Layout
+            switch self {
+            case .view(let view):
+                layout = UIViewLayout(view: view)
+            case .layer(let layer):
+                layout = CALayerLayout(layer: layer)
+            }
+            objc_setAssociatedObject(self, &layoutAssociationKey, layout, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+            return layout
+        }
+    }
+}
+
+extension ViewType: Equatable {
+    public static func == (lhs: ViewType, rhs: ViewType) -> Bool {
+        switch (lhs, rhs) {
+        case (.view(let lhs), .view(let rhs)):
+            return lhs == rhs
+        case (.layer(let lhs), .layer(let rhs)):
+            return lhs == rhs
+        default:
+            return false
+        }
+    }
+
+    public static func === (lhs: ViewType, rhs: ViewType) -> Bool {
+        switch (lhs, rhs) {
+        case (.view(let lhs), .view(let rhs)):
+            return lhs === rhs
+        case (.layer(let lhs), .layer(let rhs)):
+            return lhs === rhs
+        default:
+            return false
+        }
+    }
+}
+
+protocol Layout: class {
+    var superview: ViewType? { get }
+    var subviews: [ViewType] { get }
+    var frame: CGRect { get set}
+    var bounds: CGRect { get set}
+    var cornerRadius: CGFloat { get set}
+
+    func contains(_ view: ViewType) -> Bool
+    func convert(_ rect: CGRect, from view: ViewType?) -> CGRect
+}
+
+extension Layout {
+    func contains(_ view: ViewType) -> Bool {
+        if subviews.contains(view) {
+            return true
+        }
+        for subview in subviews {
+            if subview.layout.contains(view) {
+                return true
+            }
+        }
+        return false
+    }
+}
+
+final class UIViewLayout: Layout {
+    unowned let view: UIView
+
+    var nx_state: AnyHashable {
+        get {
+            view.nx_state
+        }
+        set {
+            view.nx_state = newValue
+        }
+    }
+
+    var superview: ViewType? {
+        guard let superview = view.superview else {
+            return nil
+        }
+        return .view(superview)
+    }
+
+    var subviews: [ViewType] {
+        return view.subviews.map { view in
+            .view(view)
+        }
+    }
+
+    var frame: CGRect {
+        get {
+            view.frame
+        }
+        set {
+            view.frame = newValue
+        }
+    }
+
+    var bounds: CGRect {
+        get {
+            view.bounds
+        }
+        set {
+            view.bounds = newValue
+        }
+    }
+
+    var cornerRadius: CGFloat {
+        get {
+            view.layer.cornerRadius
```

Comment:
Why do you insist on having duplicate logic for views and frames? This layout should be a subclass of LayerLayout.